File: flutter_app/lib/screens/quote_detail_screen.dart
import 'package:flutter/material.dart';

import 'package:flutter_app/models/quote.dart';
import 'package:flutter_app/services/quotes_service.dart';
import 'package:flutter_app/widgets/star_rating_widget.dart';

class QuoteDetailScreen extends StatefulWidget {
  final Quote quote;

  const QuoteDetailScreen({super.key, required this.quote});

  @override
  State<QuoteDetailScreen> createState() => _QuoteDetailScreenState();
}

class _QuoteDetailScreenState extends State<QuoteDetailScreen> {
  final QuotesService _quotesService = QuotesService();
  final _formKey = GlobalKey<FormState>();
  final _textController = TextEditingController();
  final _movieController = TextEditingController();
  final _characterController = TextEditingController();
  final _yearController = TextEditingController();
  late Quote _currentQuote;
  int _rating = 1; // Default rating

  @override
  void initState() {
    super.initState();
    _currentQuote = widget.quote;
    // Initialize controllers with current quote data
    _textController.text = _currentQuote.text;
    _movieController.text = _currentQuote.movie;
    _characterController.text = _currentQuote.character ?? '';
    _yearController.text = _currentQuote.year?.toString() ?? '';
    _rating = _currentQuote.rating;
  }

  @override
  void dispose() {
    _textController.dispose();
    _movieController.dispose();
    _characterController.dispose();
    _yearController.dispose();
    super.dispose();
  }

  void _updateRating(int rating) {
    setState(() {
      _rating = rating;
    });
  }

  Future<void> _saveChanges() async {
    if (_formKey.currentState!.validate()) {
      try {
        final updatedQuote = _currentQuote.copyWith(
          text: _textController.text,
          movie: _movieController.text,
          character: _characterController.text.isEmpty
              ? null
              : _characterController.text,
          year: _yearController.text.isEmpty
              ? null
              : int.tryParse(_yearController.text),
          rating: _rating,
        );

        await _quotesService.updateQuote(updatedQuote);

        setState(() {
          _currentQuote = updatedQuote;
        });

        if (mounted) {
          ScaffoldMessenger.of(context).showSnackBar(
            const SnackBar(
              content: Text('Änderungen gespeichert'),
              backgroundColor: Colors.green,
            ),
          );
          Navigator.of(context).pop(updatedQuote);
        }
      } catch (e) {
        if (mounted) {
          ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(
              content: Text('Fehler beim Speichern: $e'),
              backgroundColor: Colors.red,
            ),
          );
        }
      }
    }
  }

  Future<void> _deleteQuote() async {
    final confirmed = await showDialog<bool>(
      context: context,
      builder: (context) => AlertDialog(
        title: const Text('Zitat löschen'),
        content: const Text('Möchten Sie dieses Zitat wirklich löschen?'),
        actions: [
          TextButton(
            onPressed: () => Navigator.of(context).pop(false),
            child: const Text('Abbrechen'),
          ),
          ElevatedButton(
            onPressed: () => Navigator.of(context).pop(true),
            style: ElevatedButton.styleFrom(backgroundColor: Colors.red),
            child: const Text('Löschen', style: TextStyle(color: Colors.white)),
          ),
        ],
      ),
    );

    if (confirmed == true) {
      try {
        await _quotesService.deleteQuote(_currentQuote.id!);
        if (mounted) {
          ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(
              content: Text('Zitat gelöscht'),
              backgroundColor: Theme.of(context).colorScheme.error,
            ),
          );
          Navigator.of(context).pop('deleted');
        }
      } catch (e) {
        if (mounted) {
          ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(
              content: Text('Fehler beim Löschen: $e'),
              backgroundColor: Theme.of(context).colorScheme.error,
            ),
          );
        }
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Zitat Details')),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(24),
        child: Form(
          key: _formKey,
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              // Quote text field
              TextFormField(
                controller: _textController,
                decoration: const InputDecoration(
                  labelText: 'Zitat*',
                  hintText: 'Geben Sie den Zitat-Text ein',
                  border: OutlineInputBorder(),
                ),
                validator: (value) {
                  if (value == null || value.isEmpty) {
                    return 'Bitte geben Sie den Zitat-Text ein';
                  }
                  return null;
                },
                maxLines: 3,
              ),
              const SizedBox(height: 16),

              // Character field
              TextFormField(
                controller: _characterController,
                decoration: const InputDecoration(
                  labelText: 'Charakter',
                  hintText: 'Wer hat das Zitat gesagt?',
                  border: OutlineInputBorder(),
                ),
              ),
              const SizedBox(height: 16),

              // Movie field
              TextFormField(
                controller: _movieController,
                decoration: const InputDecoration(
                  labelText: 'Film/Serie*',
                  hintText: 'Name des Films oder der Serie',
                  border: OutlineInputBorder(),
                ),
                validator: (value) {
                  if (value == null || value.isEmpty) {
                    return 'Bitte geben Sie den Namen des Films ein';
                  }
                  return null;
                },
              ),
              const SizedBox(height: 16),

              // Year field
              TextFormField(
                controller: _yearController,
                decoration: const InputDecoration(
                  labelText: 'Jahr',
                  hintText: 'z.B. 2004',
                  border: OutlineInputBorder(),
                ),
                keyboardType: TextInputType.number,
              ),
              const SizedBox(height: 32),

              // Rating section
              const Text(
                'Zitat bewerten:',
                style: TextStyle(
                  fontSize: 16,
                  fontWeight: FontWeight.w600,
                  color: Colors.black87,
                ),
              ),
              const SizedBox(height: 12),

              // Star rating widget
              StarRatingWidget(
                rating: _rating,
                onRatingChanged: _updateRating,
                isEditable: true,
              ),
              const SizedBox(height: 48),

              // Action buttons
              Column(
                children: [
                  // Save changes button
                  SizedBox(
                    width: double.infinity,
                    child: ElevatedButton.icon(
                      onPressed: _saveChanges,
                      icon: const Icon(Icons.save, color: Colors.white),
                      label: const Text(
                        'Änderungen speichern',
                        style: TextStyle(color: Colors.white),
                      ),
                      style: ElevatedButton.styleFrom(
                        backgroundColor: Theme.of(context).colorScheme.primary,
                        padding: const EdgeInsets.symmetric(vertical: 16),
                        shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(8),
                        ),
                      ),
                    ),
                  ),
                  const SizedBox(height: 16),

                  // Delete quote button
                  SizedBox(
                    width: double.infinity,
                    child: ElevatedButton.icon(
                      onPressed: _deleteQuote,
                      icon: const Icon(Icons.delete, color: Colors.white),
                      label: const Text(
                        'Zitat löschen',
                        style: TextStyle(color: Colors.white),
                      ),
                      style: ElevatedButton.styleFrom(
                        backgroundColor: Theme.of(context).colorScheme.error,
                        padding: const EdgeInsets.symmetric(vertical: 16),
                        shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(8),
                        ),
                      ),
                    ),
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}

File: kmp_app/composeApp/src/commonMain/kotlin/org/example/project/screens/QuoteDetailScreen.kt
package org.example.project.screens

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.Save
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.FocusDirection
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch
import org.example.project.models.Quote
import org.example.project.services.QuotesService
import org.example.project.widgets.StarRating

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun QuoteDetailScreen(
    quote: Quote,
    quotesService: QuotesService,
    onBack: () -> Unit,
    onQuoteUpdated: () -> Unit
) {
    var text by remember { mutableStateOf(quote.text) }
    var movie by remember { mutableStateOf(quote.movie) }
    var character by remember { mutableStateOf(quote.character ?: "") }
    var year by remember { mutableStateOf(quote.year?.toString() ?: "") }
    var rating by remember { mutableStateOf(quote.rating) }

    var textError by remember { mutableStateOf<String?>(null) }
    var movieError by remember { mutableStateOf<String?>(null) }

    var showDeleteDialog by remember { mutableStateOf(false) }
    var isLoading by remember { mutableStateOf(false) }

    val scope = rememberCoroutineScope()
    val snackbarHostState = remember { SnackbarHostState() }
    val focusManager = LocalFocusManager.current

    fun validateForm(): Boolean {
        textError = if (text.isBlank()) "Bitte geben Sie den Zitat-Text ein" else null
        movieError = if (movie.isBlank()) "Bitte geben Sie den Namen des Films ein" else null

        return textError == null && movieError == null
    }

    fun saveChanges() {
        if (validateForm()) {
            scope.launch {
                try {
                    isLoading = true
                    val updatedQuote = quote.copy(
                        text = text,
                        movie = movie,
                        character = character.ifBlank { null },
                        year = year.toIntOrNull(),
                        rating = rating
                    )
                    quotesService.updateQuote(updatedQuote)
                    snackbarHostState.showSnackbar(
                        message = "Änderungen gespeichert",
                        duration = SnackbarDuration.Short
                    )
                    onQuoteUpdated()
                } catch (e: Exception) {
                    snackbarHostState.showSnackbar(
                        message = "Fehler beim Speichern: ${e.message}",
                        duration = SnackbarDuration.Long
                    )
                } finally {
                    isLoading = false
                }
            }
        }
    }

    fun deleteQuote() {
        scope.launch {
            try {
                isLoading = true
                quotesService.deleteQuote(quote.id!!)
                snackbarHostState.showSnackbar(
                    message = "Zitat gelöscht",
                    duration = SnackbarDuration.Short
                )
                onQuoteUpdated()
                onBack()
            } catch (e: Exception) {
                snackbarHostState.showSnackbar(
                    message = "Fehler beim Löschen: ${e.message}",
                    duration = SnackbarDuration.Long
                )
            } finally {
                isLoading = false
            }
        }
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Zitat Details") },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.Default.ArrowBack, contentDescription = "Zurück")
                    }
                }
            )
        },
        snackbarHost = { SnackbarHost(snackbarHostState) }
    ) { paddingValues ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
                .padding(24.dp)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            OutlinedTextField(
                value = text,
                onValueChange = {
                    text = it
                    textError = null
                },
                label = { Text("Zitat*") },
                placeholder = { Text("Geben Sie den Zitat-Text ein") },
                isError = textError != null,
                supportingText = textError?.let { { Text(it) } },
                modifier = Modifier.fillMaxWidth(),
                maxLines = 3,
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Next),
                keyboardActions = KeyboardActions(
                    onNext = { focusManager.moveFocus(FocusDirection.Down) }
                )
            )

            OutlinedTextField(
                value = character,
                onValueChange = { character = it },
                label = { Text("Charakter") },
                placeholder = { Text("Wer hat das Zitat gesagt?") },
                modifier = Modifier.fillMaxWidth(),
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Next),
                keyboardActions = KeyboardActions(
                    onNext = { focusManager.moveFocus(FocusDirection.Down) }
                )
            )

            OutlinedTextField(
                value = movie,
                onValueChange = {
                    movie = it
                    movieError = null
                },
                label = { Text("Film/Serie*") },
                placeholder = { Text("Name des Films oder der Serie") },
                isError = movieError != null,
                supportingText = movieError?.let { { Text(it) } },
                modifier = Modifier.fillMaxWidth(),
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Next),
                keyboardActions = KeyboardActions(
                    onNext = { focusManager.moveFocus(FocusDirection.Down) }
                )
            )

            OutlinedTextField(
                value = year,
                onValueChange = { year = it },
                label = { Text("Jahr") },
                placeholder = { Text("z.B. 2004") },
                keyboardOptions = KeyboardOptions(
                    keyboardType = KeyboardType.Number,
                    imeAction = ImeAction.Done
                ),
                keyboardActions = KeyboardActions(
                    onDone = { focusManager.clearFocus() }
                ),
                modifier = Modifier.fillMaxWidth()
            )

            Spacer(modifier = Modifier.height(16.dp))

            Text(
                text = "Zitat bewerten:",
                fontSize = 16.sp,
                fontWeight = FontWeight.SemiBold,
                color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.87f)
            )

            StarRating(
                rating = rating,
                starSize = 32.dp,
                isEditable = true,
                onRatingChanged = { rating = it }
            )

            Spacer(modifier = Modifier.height(32.dp))

            Button(
                onClick = { saveChanges() },
                modifier = Modifier.fillMaxWidth(),
                enabled = !isLoading
            ) {
                if (isLoading) {
                    CircularProgressIndicator(
                        modifier = Modifier.size(16.dp),
                        color = Color.White
                    )
                } else {
                    Icon(Icons.Default.Save, contentDescription = null)
                    Spacer(modifier = Modifier.width(8.dp))
                    Text("Änderungen speichern")
                }
            }

            OutlinedButton(
                onClick = { showDeleteDialog = true },
                modifier = Modifier.fillMaxWidth(),
                enabled = !isLoading,
                colors = ButtonDefaults.outlinedButtonColors(
                    contentColor = MaterialTheme.colorScheme.error
                )
            ) {
                Icon(Icons.Default.Delete, contentDescription = null)
                Spacer(modifier = Modifier.width(8.dp))
                Text("Zitat löschen")
            }
        }
    }

    if (showDeleteDialog) {
        AlertDialog(
            onDismissRequest = { showDeleteDialog = false },
            title = { Text("Zitat löschen") },
            text = { Text("Möchten Sie dieses Zitat wirklich löschen?") },
            confirmButton = {
                Button(
                    onClick = {
                        showDeleteDialog = false
                        deleteQuote()
                    },
                    colors = ButtonDefaults.buttonColors(
                        containerColor = MaterialTheme.colorScheme.error
                    )
                ) {
                    Text("Löschen", color = Color.White)
                }
            },
            dismissButton = {
                TextButton(onClick = { showDeleteDialog = false }) {
                    Text("Abbrechen")
                }
            }
        )
    }
}
